iwyu_output.cc

```diff
@@ -72,53 +72,60 @@ class OutputLine {
+    symbols_.erase(std::remove_if(symbols_.begin(), symbols_.end(),
+                                  [](const string& v) { return v.empty(); }),
+                   symbols_.end());
+  }
 
   size_t line_length() const { return line_.size(); }
   bool needs_alignment() const { return !symbols_.empty(); }
   void add_prefix(const string& prefix) { line_ = prefix + line_; }
-  string printable_line(size_t desired_length, size_t max_length) const;
+  string printable_line(size_t min_length, size_t max_length) const;
 
  private:
   string line_;                     // '#include XXX' or 'class YYY;'
   vector<string> symbols_;          // symbols used from included header
 };
 
-string OutputLine::printable_line(size_t desired_length, size_t max_length)
-    const {
-  // Reduce max_length to make room for the ", etc" suffix.
-  max_length -= strlen(", etc");
-
-  desired_length = std::min(desired_length, max_length);
-
-  string retval = line_;
-
-  string prefix;   // what we print before each symbol in the 'why' comments
-  // We try to get the columns to line up nicely.
-  if (retval.length() < desired_length)
-    prefix += string(desired_length - retval.length(), ' ');
-  prefix += "  // for ";    // before 1st symbol, print ' // for '
-  int symbols_printed = 0;
-
-  for (Each<string> it(&symbols_); !it.AtEnd(); ++it) {
-    if (it->empty())       // ignore the empty ("") symbol
-      continue;
-    // At verbose levels of 0, 1, or 2, cut off output at max_length columns.
-    // Actually, at 74, to leave 5 chars for ', etc' and 1 for newline.
-    if (ShouldPrint(3) ||
-        retval.length() + prefix.length() + it->length() <= max_length) {
-      retval += prefix + *it;
-      ++symbols_printed;
-      prefix = ", ";       // before 2nd and subsequent symbols, print ', '
-    } else {
-      // Truncate at max_length cols.
-      if (symbols_printed > 0)
-        retval += ", etc";
-      break;
+// Append a helpful 'why' comment to the include line, containing the symbols
+// used from the header. Align nicely at lower verbosity levels.
+string OutputLine::printable_line(size_t min_length, size_t max_length) const {
+  // If there are no symbols to mention, return the line as-is.
+  if (symbols_.empty())
+    return line_;
+
+  --max_length;  // Spare a char for newline character.
+
+  // Before first symbol, print '  // for ' and pad it so 'why' comments are
+  // nicely aligned.
+  string symbol_prefix = "  // for ";
+  if (min_length > line_.length())
+    symbol_prefix.insert(0, min_length - line_.length(), ' ');
+
+  string result = line_;
+  for (string symbol : symbols_) {
+    // At verbose levels 0-2, truncate output to max_length columns.
+    if (!ShouldPrint(3)) {
+      // Calculate number of chars remaining.
+      size_t remaining = 0;
+      size_t result_length = result.length() + symbol_prefix.length();
+      if (max_length > result_length)
+        remaining = max_length - result_length;
+
```

Comment:
And here too, I prefer `result_length < max_length`. Interesting, if there is statistic data regarding usage of `<` and `>`.